src/Juvix/Compiler/Core/Transformation/LambdaLifting.hs

```diff
@@ -4,13 +4,36 @@ module Juvix.Compiler.Core.Transformation.LambdaLifting
   )
 where
 
+import Juvix.Compiler.Core.Data.BinderList (BinderList)
+import Juvix.Compiler.Core.Data.BinderList qualified as BL
 import Juvix.Compiler.Core.Data.InfoTableBuilder
+import Juvix.Compiler.Core.Extra
+import Juvix.Compiler.Core.Pretty
 import Juvix.Compiler.Core.Transformation.Base
 
-lambdaLiftNode :: Member InfoTableBuilder r => Node -> Sem r Node
-lambdaLiftNode _ = do
-  void freshSymbol
-  error "not yet implemented"
+lambdaLiftNode :: forall r. Member InfoTableBuilder r => BinderList Info -> Node -> Sem r Node
+lambdaLiftNode aboveBl top = mkLambdas topArgs <$> dmapLRM' (topArgsBinderList <> aboveBl, go) body
+  where
+    (topArgs, body) = unfoldLambdas' top
+    topArgsBinderList :: BinderList Info
+    topArgsBinderList = BL.fromList topArgs
+    go :: BinderList Info -> Node -> Sem r Recur
+    go bl = \case
+      l@Lambda {} -> do
+        l' <- lambdaLiftNode bl l
+        let freevars = toList (getFreeVars l')
+            freevarsAssocs = [(i, BL.lookup i bl) | i <- freevars]
+            fBody' = captureFreeVars freevarsAssocs l'
+        f <- freshSymbol
+        traceM "register"
+        registerIdentNode f fBody'
```

Comment:
I forgot, but we should also be creating an IdentifierInfo for each new function. This can be done with registerIdent from InfoTableBuilder effect. Probably we should change the name fields in IdentifierInfo, InductiveInfo, etc., to Maybe types. The `Type` fields should be filled in as faithfully as possible, but one can always use Dynamic so Maybe is never necessary.

Reply:
Do you think it's better to add fetching the type information in a separate PR? We could also do it that way, but it's important to have it ultimately.

For now, one could safely assume that the Info in the BinderList for a given variable contains TypeInfo which gives the type of the variable (or use Dynamic if the TypeInfo is not there). So at this stage one could easily add type information for function arguments. For the result, it could be decided how to do it after the translation Internal -> Core is complete.

Reply:
Unless I'm missing something, that's what I'm doing in `typeFromArgs`

Reply:
Sorry, I somehow missed that. I didn't look too closely. Since you didn't reply to the original comment I assumed it wasn't fixed.

Reply:
I think it's fine now.